# Project # 4 – Predictive Analysis using scikit-learn
 
### Your assignment is to:


- Start with the mushroom data in the pandas DataFrame that you constructed in your “Assignment – Preprocessing Data with sci-kit learn.”


- Use scikit-learn to determine which of the two predictor columns that you selected (odor and one other column of your choice) most accurately predicts whether or not a mushroom is poisonous. There is an additional challenge here—to use scikit-learn’s predictive classifiers, you’ll want to convert each of your two (numeric categorical) predictor columns into a set of columns. See for one approach pandas get_dummies() method.


- Clearly state your conclusions along with any recommendations for further analysis.

This is by design a very open-ended assignment. You’re encouraged to go through the resources in your Week 12 folder. In particular, if you understand the process used in the Kevin Markham videos on Machine Learning with scikit-learn to predict iris species from four predictor variables, you should be able to transfer your learnings to complete this task.

***
First lets import libraries we need and the data from week 13 assignment

In [1]:

import pandas as pd
from matplotlib import pyplot as plt
import seaborn as sns
%matplotlib inline
import numpy as np


from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import sklearn.model_selection
from sklearn import metrics

In [2]:
import_file = 'agaricus-lepiota.data'
mushroom_dataset = pd.read_csv(import_file, header=None, usecols=[0, 3, 5])
mushroom_dataset.head(11)

0  3  5
0   p  n  p
1   e  y  a
2   e  w  l
3   p  w  p
4   e  g  n
5   e  y  a
6   e  w  a
7   e  w  l
8   p  w  p
9   e  y  a
10  e  y  l

In [3]:
column_names = ['Edible/Poisonous', 'Cap-Color', 'Odor']
mushroom_dataset.columns = column_names
mushroom_dataset.head(11)

Edible/Poisonous Cap-Color Odor
0                 p         n    p
1                 e         y    a
2                 e         w    l
3                 p         w    p
4                 e         g    n
5                 e         y    a
6                 e         w    a
7                 e         w    l
8                 p         w    p
9                 e         y    a
10                e         y    l

In [4]:
replace_with_num_values = mushroom_dataset.replace({'Edible/Poisonous':{'e':0,
                                                                        'p':1},
                                                    'Cap-Color':{'n':0,
                                                                 'b':1,
                                                                 'c':2,
                                                                 'g':3,
                                                                 'r':4,
                                                                 'p':5,
                                                                 'u':6,
                                                                 'e':7,
                                                                 'w':8,
                                                                 'y':9},
                                                    'Odor':{'a': 0,
                                                            'l': 1,
                                                            'c': 2,
                                                            'y': 3,
                                                            'f': 4,
                                                            'm': 5,
                                                            'n': 6,
                                                            'p': 7,
                                                            's': 8}})
replace_with_num_values.head(11)

Edible/Poisonous  Cap-Color  Odor
0                  1          0     7
1                  0          9     0
2                  0          8     1
3                  1          8     7
4                  0          3     6
5                  0          9     0
6                  0          8     0
7                  0          8     1
8                  1          8     7
9                  0          9     0
10                 0          9     1

# I found it esier to change the values to be unique to each column so they can be serched for using the `.iloc` later during our testing

In [5]:


replace_with_num_values = mushroom_dataset.replace({'Edible/Poisonous':{'e':0,
                                                                        'p':1},
                                                    'Cap-Color':{'n':2,
                                                                 'b':3,
                                                                 'c':4,
                                                                 'g':5,
                                                                 'r':6,
                                                                 'p':7,
                                                                 'u':8,
                                                                 'e':9,
                                                                 'w':10,
                                                                 'y':11},
                                                    'Odor':{'a': 12,
                                                            'l': 13,
                                                            'c': 14,
                                                            'y': 15,
                                                            'f': 16,
                                                            'm': 17,
                                                            'n': 18,
                                                            'p': 19,
                                                            's': 20}})
replace_with_num_values.head(11)

Edible/Poisonous  Cap-Color  Odor
0                  1          2    19
1                  0         11    12
2                  0         10    13
3                  1         10    19
4                  0          5    18
5                  0         11    12
6                  0         10    12
7                  0         10    13
8                  1         10    19
9                  0         11    12
10                 0         11    13

***
Now we will create our `getdummies()` data

In [18]:
odor_series = pd.Series(replace_with_num_values['Odor'])
dummy_odor = pd.get_dummies(odor_series)
dummy_odor

12  13  14  15  16  17  18  19  20
0      0   0   0   0   0   0   0   1   0
1      1   0   0   0   0   0   0   0   0
2      0   1   0   0   0   0   0   0   0
3      0   0   0   0   0   0   0   1   0
4      0   0   0   0   0   0   1   0   0
5      1   0   0   0   0   0   0   0   0
6      1   0   0   0   0   0   0   0   0
7      0   1   0   0   0   0   0   0   0
8      0   0   0   0   0   0   0   1   0
9      1   0   0   0   0   0   0   0   0
10     0   1   0   0   0   0   0   0   0
11     1   0   0   0   0   0   0   0   0
12     1   0   0   0   0   0   0   0   0
13     0   0   0   0   0   0   0   1   0
14     0   0   0   0   0   0   1   0   0
15     0   0   0   0   0   0   1   0   0
16     0   0   0   0   0   0   1   0   0
17     0   0   0   0   0   0   0   1   0
18     0   0   0   0   0   0   0   1   0
19     0   0   0   0   0   0   0   1   0
20     1   0   0   0   0   0   0   0   0
21     0   0   0   0   0   0   0   1   0
22     0   1   0   0   0   0   0   0   0
23     1   0   0   0   0   0   0   0   0
24     0   1   0   0   0   0   0   0   0
25     0   0   0   0   0   0   0   1   0
26     1   0   0   0   0   0   0   0   0
27     0   1   0   0   0   0   0   0   0
28     0   0   0   0   0   0   1   0   0
29     1   0   0   0   0   0   0   0   0
...   ..  ..  ..  ..  ..  ..  ..  ..  ..
8094   0   0   0   0   0   0   1   0   0
8095   0   0   0   0   0   1   0   0   0
8096   0   0   0   0   0   0   1   0   0
8097   0   0   0   0   0   0   0   0   1
8098   0   0   0   1   0   0   0   0   0
8099   0   0   0   0   0   0   1   0   0
8100   0   0   0   0   0   0   1   0   0
8101   0   0   0   0   0   0   0   0   1
8102   0   0   0   0   0   0   1   0   0
8103   0   0   0   0   0   0   1   0   0
8104   0   0   0   0   0   0   1   0   0
8105   0   0   0   0   0   0   1   0   0
8106   0   0   0   0   0   0   1   0   0
8107   0   0   0   0   0   0   1   0   0
8108   0   0   0   1   0   0   0   0   0
8109   0   0   0   0   0   0   1   0   0
8110   0   0   0   0   0   0   1   0   0
8111   0   0   0   0   0   0   1   0   0
8112   0   0   0   0   0   0   1   0   0
8113   0   0   0   1   0   0   0   0   0
8114   0   0   0   0   0   1   0   0   0
8115   0   0   0   0   0   0   1   0   0
8116   0   0   0   0   0   0   0   0   1
8117   0   0   0   1   0   0   0   0   0
8118   0   0   0   0   1   0   0   0   0
8119   0   0   0   0   0   0   1   0   0
8120   0   0   0   0   0   0   1   0   0
8121   0   0   0   0   0   0   1   0   0
8122   0   0   0   1   0   0   0   0   0
8123   0   0   0   0   0   0   1   0   0

[8124 rows x 9 columns]

In [19]:
cap_color_series = pd.Series(replace_with_num_values['Cap-Color'])
dummy_cap_color = pd.get_dummies(cap_color_series)
dummy_cap_color

2   3   4   5   6   7   8   9   10  11
0      1   0   0   0   0   0   0   0   0   0
1      0   0   0   0   0   0   0   0   0   1
2      0   0   0   0   0   0   0   0   1   0
3      0   0   0   0   0   0   0   0   1   0
4      0   0   0   1   0   0   0   0   0   0
5      0   0   0   0   0   0   0   0   0   1
6      0   0   0   0   0   0   0   0   1   0
7      0   0   0   0   0   0   0   0   1   0
8      0   0   0   0   0   0   0   0   1   0
9      0   0   0   0   0   0   0   0   0   1
10     0   0   0   0   0   0   0   0   0   1
11     0   0   0   0   0   0   0   0   0   1
12     0   0   0   0   0   0   0   0   0   1
13     0   0   0   0   0   0   0   0   1   0
14     1   0   0   0   0   0   0   0   0   0
15     0   0   0   1   0   0   0   0   0   0
16     0   0   0   0   0   0   0   0   1   0
17     1   0   0   0   0   0   0   0   0   0
18     0   0   0   0   0   0   0   0   1   0
19     1   0   0   0   0   0   0   0   0   0
20     0   0   0   0   0   0   0   0   0   1
21     1   0   0   0   0   0   0   0   0   0
22     0   0   0   0   0   0   0   0   0   1
23     0   0   0   0   0   0   0   0   1   0
24     0   0   0   0   0   0   0   0   1   0
25     0   0   0   0   0   0   0   0   1   0
26     0   0   0   0   0   0   0   0   0   1
27     0   0   0   0   0   0   0   0   1   0
28     1   0   0   0   0   0   0   0   0   0
29     0   0   0   0   0   0   0   0   0   1
...   ..  ..  ..  ..  ..  ..  ..  ..  ..  ..
8094   0   0   0   1   0   0   0   0   0   0
8095   0   0   1   0   0   0   0   0   0   0
8096   0   0   0   0   0   0   0   0   1   0
8097   1   0   0   0   0   0   0   0   0   0
8098   0   0   0   0   0   0   0   1   0   0
8099   0   0   0   0   0   0   0   0   1   0
8100   1   0   0   0   0   0   0   0   0   0
8101   0   0   0   0   0   0   0   1   0   0
8102   1   0   0   0   0   0   0   0   0   0
8103   1   0   0   0   0   0   0   0   0   0
8104   1   0   0   0   0   0   0   0   0   0
8105   1   0   0   0   0   0   0   0   0   0
8106   1   0   0   0   0   0   0   0   0   0
8107   1   0   0   0   0   0   0   0   0   0
8108   0   0   0   0   0   0   0   1   0   0
8109   0   0   0   0   0   0   0   0   1   0
8110   1   0   0   0   0   0   0   0   0   0
8111   0   0   0   0   0   0   0   0   1   0
8112   1   0   0   0   0   0   0   0   0   0
8113   0   0   0   0   0   0   0   1   0   0
8114   0   0   1   0   0   0   0   0   0   0
8115   1   0   0   0   0   0   0   0   0   0
8116   1   0   0   0   0   0   0   0   0   0
8117   0   0   0   0   0   0   0   1   0   0
8118   1   0   0   0   0   0   0   0   0   0
8119   1   0   0   0   0   0   0   0   0   0
8120   1   0   0   0   0   0   0   0   0   0
8121   1   0   0   0   0   0   0   0   0   0
8122   1   0   0   0   0   0   0   0   0   0
8123   1   0   0   0   0   0   0   0   0   0

[8124 rows x 10 columns]

Now we will create a new data frame by merging our dummy columns to create `Mushroom_dummy_dataframe`

In [8]:
Mushroom_dummy_dataframe = pd.concat([dummy_odor, dummy_cap_color, replace_with_num_values['Edible/Poisonous']], axis=1)
Mushroom_dummy_dataframe

12  13  14  15  16  17  18  19  20  2  3  4  5  6  7  8  9  10  11  \
0      0   0   0   0   0   0   0   1   0  1  0  0  0  0  0  0  0   0   0   
1      1   0   0   0   0   0   0   0   0  0  0  0  0  0  0  0  0   0   1   
2      0   1   0   0   0   0   0   0   0  0  0  0  0  0  0  0  0   1   0   
3      0   0   0   0   0   0   0   1   0  0  0  0  0  0  0  0  0   1   0   
4      0   0   0   0   0   0   1   0   0  0  0  0  1  0  0  0  0   0   0   
5      1   0   0   0   0   0   0   0   0  0  0  0  0  0  0  0  0   0   1   
6      1   0   0   0   0   0   0   0   0  0  0  0  0  0  0  0  0   1   0   
7      0   1   0   0   0   0   0   0   0  0  0  0  0  0  0  0  0   1   0   
8      0   0   0   0   0   0   0   1   0  0  0  0  0  0  0  0  0   1   0   
9      1   0   0   0   0   0   0   0   0  0  0  0  0  0  0  0  0   0   1   
10     0   1   0   0   0   0   0   0   0  0  0  0  0  0  0  0  0   0   1   
11     1   0   0   0   0   0   0   0   0  0  0  0  0  0  0  0  0   0   1   
12     1   0   0   0   0   0   0   0   0  0  0  0  0  0  0  0  0   0   1   
13     0   0   0   0   0   0   0   1   0  0  0  0  0  0  0  0  0   1   0   
14     0   0   0   0   0   0   1   0   0  1  0  0  0  0  0  0  0   0   0   
15     0   0   0   0   0   0   1   0   0  0  0  0  1  0  0  0  0   0   0   
16     0   0   0   0   0   0   1   0   0  0  0  0  0  0  0  0  0   1   0   
17     0   0   0   0   0   0   0   1   0  1  0  0  0  0  0  0  0   0   0   
18     0   0   0   0   0   0   0   1   0  0  0  0  0  0  0  0  0   1   0   
19     0   0   0   0   0   0   0   1   0  1  0  0  0  0  0  0  0   0   0   
20     1   0   0   0   0   0   0   0   0  0  0  0  0  0  0  0  0   0   1   
21     0   0   0   0   0   0   0   1   0  1  0  0  0  0  0  0  0   0   0   
22     0   1   0   0   0   0   0   0   0  0  0  0  0  0  0  0  0   0   1   
23     1   0   0   0   0   0   0   0   0  0  0  0  0  0  0  0  0   1   0   
24     0   1   0   0   0   0   0   0   0  0  0  0  0  0  0  0  0   1   0   
25     0   0   0   0   0   0   0   1   0  0  0  0  0  0  0  0  0   1   0   
26     1   0   0   0   0   0   0   0   0  0  0  0  0  0  0  0  0   0   1   
27     0   1   0   0   0   0   0   0   0  0  0  0  0  0  0  0  0   1   0   
28     0   0   0   0   0   0   1   0   0  1  0  0  0  0  0  0  0   0   0   
29     1   0   0   0   0   0   0   0   0  0  0  0  0  0  0  0  0   0   1   
...   ..  ..  ..  ..  ..  ..  ..  ..  .. .. .. .. .. .. .. .. ..  ..  ..   
8094   0   0   0   0   0   0   1   0   0  0  0  0  1  0  0  0  0   0   0   
8095   0   0   0   0   0   1   0   0   0  0  0  1  0  0  0  0  0   0   0   
8096   0   0   0   0   0   0   1   0   0  0  0  0  0  0  0  0  0   1   0   
8097   0   0   0   0   0   0   0   0   1  1  0  0  0  0  0  0  0   0   0   
8098   0   0   0   1   0   0   0   0   0  0  0  0  0  0  0  0  1   0   0   
8099   0   0   0   0   0   0   1   0   0  0  0  0  0  0  0  0  0   1   0   
8100   0   0   0   0   0   0   1   0   0  1  0  0  0  0  0  0  0   0   0   
8101   0   0   0   0   0   0   0   0   1  0  0  0  0  0  0  0  1   0   0   
8102   0   0   0   0   0   0   1   0   0  1  0  0  0  0  0  0  0   0   0   
8103   0   0   0   0   0   0   1   0   0  1  0  0  0  0  0  0  0   0   0   
8104   0   0   0   0   0   0   1   0   0  1  0  0  0  0  0  0  0   0   0   
8105   0   0   0   0   0   0   1   0   0  1  0  0  0  0  0  0  0   0   0   
8106   0   0   0   0   0   0   1   0   0  1  0  0  0  0  0  0  0   0   0   
8107   0   0   0   0   0   0   1   0   0  1  0  0  0  0  0  0  0   0   0   
8108   0   0   0   1   0   0   0   0   0  0  0  0  0  0  0  0  1   0   0   
8109   0   0   0   0   0   0   1   0   0  0  0  0  0  0  0  0  0   1   0   
8110   0   0   0   0   0   0   1   0   0  1  0  0  0  0  0  0  0   0   0   
8111   0   0   0   0   0   0   1   0   0  0  0  0  0  0  0  0  0   1   0   
8112   0   0   0   0   0   0   1   0   0  1  0  0  0  0  0  0  0   0   0   
8113   0   0   0   1   0   0   0   0   0  0  0  0  0  0  0  0  1   0   0   
8114   0   0   0   0   0   1   0   0   0  0  0  1  0  

***
Lets now take a loook at our columns in `Mushroom_dummy_dataframe`

In [20]:
columns = list(Mushroom_dummy_dataframe.iloc[:, :])
columns

[12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 'Edible/Poisonous']

***

Now we need to creat our `x` and `y` for our testing

In [23]:
x = Mushroom_dummy_dataframe.iloc[:, :-1].values
x

array([[0, 0, 0, ..., 0, 0, 0],
       [1, 0, 0, ..., 0, 0, 1],
       [0, 1, 0, ..., 0, 1, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=uint8)

In [24]:
y = Mushroom_dummy_dataframe.iloc[:, 1].values
y

array([0, 0, 1, ..., 0, 0, 0], dtype=uint8)

***
Now we can create our tranind model and apply linner regression  to test data

In [27]:
x_train, x_test, y_train, y_test = train_test_split(x, y, random_state=1)

In [28]:
linreg = sklearn.linear_model.LinearRegression()
linreg.fit(x_train, y_train)
y_pred = linreg.predict(x_test)
t = [1, 0]
p = [1, 0]

***
Now we can apply the test to `Odor` values  to see how acurate `Odor` predicts `Edible` or `Posinuos` Mushrooms

In [34]:
x = Mushroom_dummy_dataframe.iloc[:, 12:20].values
y = Mushroom_dummy_dataframe.iloc[:, 1].values
x_train, x_test, y_train, y_test = sklearn.model_selection.train_test_split(x, y, random_state=1)
linreg.fit(x_train, y_train)
y_pred = linreg.predict(x_test)

print(np.sqrt(metrics.mean_squared_error(y_test, y_pred)))

0.1961921781160766


*** Now we can apply the test to `Cap-Color` values  to see how acurate `Cap-Color` predicts `Edible` or `Posinuos` Mushrooms

In [35]:
x = Mushroom_dummy_dataframe.iloc[:, 2:11].values
y = Mushroom_dummy_dataframe.iloc[:, 1].values
x_train, x_test, y_train, y_test = sklearn.model_selection.train_test_split(x, y, random_state=1)
linreg.fit(x_train, y_train)
y_pred = linreg.predict(x_test)

print(np.sqrt(metrics.mean_squared_error(y_test, y_pred)))

0.15654823970004056


### In conclusion, as we can see based on our test results the `Cap-Color` result of `0.15654823970004056` is closer to Zero(0) than our `Odor` result of `0.1961921781160766`, indicating that `Cap-Color` is a better indicator of ***Edible or Posinous*** than `Odor` 